File: test/benchmark/binary_trees.dart
// Ported from the Wren version.

class Tree {
  var _item;
  var _left;
  var _right;

  Tree(item, depth) {
    _item = item;
    if (depth > 0) {
      var item2 = item + item;
      depth--;
      _left = new Tree(item2 - 1, depth);
      _right = new Tree(item2, depth);
    }
  }

  get check {
    if (_left == null) {
      return _item;
    }

    return _item + _left.check - _right.check;
  }
}

main() {
  var minDepth = 4;
  var maxDepth = 12;
  var stretchDepth = maxDepth + 1;

  Stopwatch watch = new Stopwatch();
  watch.start();

  print("stretch tree of depth ${stretchDepth} check: "
      "${new Tree(0, stretchDepth).check}");

  var longLivedTree = new Tree(0, maxDepth);

  // iterations = 2 ** maxDepth
  var iterations = 1;
  for (var d = 0; d < maxDepth; d++) {
    iterations = iterations * 2;
  }

  var depth = minDepth;
  while (depth < stretchDepth) {
    var check = 0;
    for (var i = 1; i <= iterations; i++) {
      check += new Tree(i, depth).check + new Tree(-i, depth).check;
    }

    print("${iterations * 2} trees of depth ${depth} check: ${check}");
    iterations ~/= 4;
    depth += 2;
  }

  print("long lived tree of depth ${maxDepth} check: ${longLivedTree.check}");
  print("elapsed: ${watch.elapsedMilliseconds / 1000}");
}

File: test/benchmark/binary_trees.py
# The Computer Language Benchmarks Game
# http://shootout.alioth.debian.org/
#
# contributed by Antoine Pitrou
# modified by Dominique Wahli
# modified by Heinrich Acker
from __future__ import print_function

import time

# Map "range" to an efficient range in both Python 2 and 3.
try:
    range = xrange
except NameError:
    pass

def make_tree(item, depth):
    if not depth: return item, None, None
    item2 = item + item
    depth -= 1
    return item, make_tree(item2 - 1, depth), make_tree(item2, depth)

def check_tree(node):
    item, left, right = node
    if not left: return item
    return item + check_tree(left) - check_tree(right)

min_depth = 4
max_depth = 12
stretch_depth = max_depth + 1

start = time.process_time()
print("stretch tree of depth %d check:" % stretch_depth, check_tree(make_tree(0, stretch_depth)))

long_lived_tree = make_tree(0, max_depth)

iterations = 2 ** max_depth
for depth in range(min_depth, stretch_depth, 2):

    check = 0
    for i in range(1, iterations + 1):
        check += check_tree(make_tree(i, depth)) + check_tree(make_tree(-i, depth))

    print("%d trees of depth %d check:" % (iterations * 2, depth), check)
    iterations //= 4

print("long lived tree of depth %d check:" % max_depth, check_tree(long_lived_tree))
print("elapsed: " + str(time.process_time() - start))